controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/SealStreamTask.java

```diff
@@ -72,6 +91,43 @@ public SealStreamTask(final StreamMetadataTasks streamMetadataTasks,
                 });
     }
 
+    private CompletableFuture<Boolean> cancelTransactions(OperationContext context, String scope, String stream) {
+        return streamMetadataStore.getActiveTxns(scope, stream, context, executor)
+                .thenCompose(activeTxns -> {
+                    if (activeTxns == null || activeTxns.isEmpty()) {
+                        return CompletableFuture.completedFuture(true);
+                    } else {
+                        // abort transactions
+                        return Futures.allOfWithResults(activeTxns.entrySet().stream().map(txIdPair -> {
+                            CompletableFuture<Void> voidCompletableFuture;
+                            if (txIdPair.getValue().getTxnStatus().equals(TxnStatus.OPEN)) {
+                                voidCompletableFuture = Futures.toVoid(streamTransactionMetadataTasks
+                                        .abortTxn(scope, stream, txIdPair.getKey(), null, context)
+                                        .exceptionally(e -> {
+                                            Throwable cause = Exceptions.unwrap(e);
+                                            if (cause instanceof StoreException.IllegalStateException ||
+                                                    cause instanceof StoreException.WriteConflictException) {
+                                                // IllegalStateException : The transaction is already in the process of being
+                                                // completed. Ignore
+                                                // WriteConflictException : Another thread is updating the transaction record.
+                                                // ignore. We will effectively retry cleaning up the transaction if it is not
+                                                // already being aborted.
+                                                return null;
+                                            } else {
+                                                // throw the original exception
+                                                throw new CompletionException(e);
+                                            }
+                                        }));
+                            } else {
+                                voidCompletableFuture = CompletableFuture.completedFuture(null);
+                            }
+
+                            return voidCompletableFuture;
+                        }).collect(Collectors.toList())).thenApply(v -> false);
```

Comment:
Why do you collect these into a list and then ignore that just so you return `false` every time?

Reply:
The list  is passed to Futures.allOfWithResults. 

We return a boolean to let the caller if there are transactions on the stream. If this method returns `true`, the caller can then proceed with sealing of stream. `False` means the caller should get the processing of seal postponed.

Reply:
In this `else` block, you are aborting the txns, which is part of the change of semantics here, so you need to execute the logic. The thing that is confusing is why you need to collect onto a list if you don't really need it and always return false. Are you doing this to be able to wait until aborting complete for all txns?

Reply:
abortTxn method simply posts event in event processor stream after setting txn state to aborting. 

The processing of abort workflow happens asynchronously. 

What we wait for here is that abort request is successfully sent out for all outstanding transactions.. 
we dont wait for abort to complete. 
we issue abort request for all of them in parallel and wait on a combined future which tells us all abort requests have been posted.. 
after that we send out "false" to the caller indicating that it cannot proceed with the sealing because we had encountered transactions.. 
This causes processing to fail with OperationNotAllowed which results in processing being postponed and retried later by which time if transactions have aborted, the sealing can commence else the same thing is repeated.